### PR TITLE
Add case sensitivity parameter to getAccount

### PR DESCRIPTION
**Pull request by Dezash**

### CODE REVIEW

**Comment by botder** [comment]
Note for reviewer: Account names are case-sensitive by default, but addAccount supports case-insensitivity.

**Comment by ccw808** [comment]
If there are two accounts 'bob' and 'BOB', and bob is first in the account database, then getAccount("BOB", nil, true) will return the account for 'bob'

**Comment by Dezash** [comment]
No it won't.
srun addAccount("bob", "hi", true)
[15:26:23] Console executed command: addAccount("bob", "hi", true)
[15:26:23] Command results: userdata: 075A41F8 [userdata]
srun addAccount("BOB", "hey", true)
[15:26:29] Console executed command: addAccount("BOB", "hey", true)
[15:26:29] Command results: userdata: 075A4400 [userdata]
srun getAccountName(getAccount("BOB", nil, true))
[15:26:54] Console executed command: getAccountName(getAccount("BOB", nil, true))
[15:26:54] Command results: BOB [string]

If you set the third argument to false (case insensitive) then it will return the first match however. Since this is not the default behavior, I think it would be acceptable if a warning in the wiki is added. Returning a table of accounts would make the function much slower so I chose to just return the first result.

**Comment by Dezash** [comment]
I just looked into it again and noticed that it generates a an array anyway so it wouldn't make it slower. Should I make it return a table when caseSensitive is set to false?
Edit: Although I'm not so sure why would anyone use this if they had accounts with case variations.

**Comment by ccw808** [comment]
What about doing a case sensitive search first? If that returns no results then do a case insensitive search

**Comment by Dezash** [comment]
If there is no case sensitive match, should it return the first case insensitive match or all case insensitive matches?

**Comment by CrosRoad95** [comment]
but you can have `bob` and `BOB`, there should be option to get N result, by default return table with all found accounts.

**Comment by Dezash** [comment]
I think returning a table would be inconsistent. Maybe there should be another function for this or an updated getAccount**s** function that could take a username argument and find all accounts with different case variations.

**Comment by Citizen01** [comment]
Something like `findAccounts` could be introduced to prevent braking scripts by changing the returned value type of `getAccount`

**Comment by Dezash** [comment]
@Citizen01 it wouldn't break any scripts since no scripts use a third argument yet. What I mean is the function name itself screams one account, singular.

**Comment by Dezash** [comment]
What do devs think?
1. Add username argument to getAccounts
2. Implement findAccounts
3. Return a table from getAccount
4. Leave it as it is (returns exact match and if there is no exact match then the first case-insensitive match)
5. Something else???

**Comment by CrosRoad95** [comment]
`Implement findAccounts` with regex support :)

**Comment by ccw808** [comment]
_4. Leave it as it is (returns exact match and if there is no exact match then the first case-insensitive match)_

Plus the password should be checked during the search as well

**Comment by ccw808** [comment]
Just to rewind a bit, when would this feature be used by scripters?

**Comment by Dezash** [comment]
One use would be to make login panels ignore case, thus reducing the chance of people forgetting their credentials.

**Comment by qaisjp** [comment]
If `4. Leave it as it is (returns exact match and if there is no exact match then the first case-insensitive match)` is accurate, isn't what you've described already possible? As long as you prevent people from creating accounts with the same casing everything should be OK.

**Comment by Dezash** [comment]
@qaisjp yes, this was to address servers that use case variations for their account usernames.

**Comment by qaisjp** [comment]
I see. So you do want to disable the ability to accept case insensitive names because you (e.g) have accounts `James` and `JAMES`, and do not want to confuse users with the ability to log in with `james`?

**Comment by Dezash** [comment]
No, it is already like that. I want to **add** an ability to accept case insensitive usernames. If the login panel example is too confusing, consider this: you want to add a real-time username availability checking (as user types their username). If there is an account called "JAMES" and user types in "James", getAccount("James") will return false and the script will say that the username is available even though it isn't since case variation is not allowed by that server. By using the feature from this PR you can check whether any case variation of "james" exists and thus return the correct result.

File: Server/mods/deathmatch/logic/CAccountManager.cpp [hunk 1]
```diff
@@ -387,12 +387,33 @@ bool CAccountManager::IntegrityCheck()
...
+                    {
+                        return pAccount;
+                    }
+                    else if (firstMatchAccount == nullptr)
+                    {
+                        firstMatchAccount = pAccount;
+                    }
+                }
+            }
+            return firstMatchAccount != nullptr ? firstMatchAccount : NULL;
```
Comment (botder) [inline Server/mods/deathmatch/logic/CAccountManager.cpp]: Can be reduced to ` return firstMatchAccount;`

File: Server/mods/deathmatch/logic/CAccountManager.cpp [hunk 2]
```diff
@@ -387,12 +387,33 @@ bool CAccountManager::IntegrityCheck()
...
 {
     if (szName && szName[0])
     {
         std::vector<CAccount*> results;
-        m_List.FindAccountMatches(&results, szName, true);
+        m_List.FindAccountMatches(&results, szName, bCaseSensitive);
+        if (!bCaseSensitive)
+        {
+            CAccount* firstMatchAccount = nullptr;
+            for (uint i = 0; i < results.size(); i++)
```
Comment (botder) [inline Server/mods/deathmatch/logic/CAccountManager.cpp]: Use a range-for loop: `for (CAccount* pAccount : results)`

File: Server/mods/deathmatch/logic/CAccountManager.cpp [hunk 3]
```diff
@@ -387,12 +387,33 @@ bool CAccountManager::IntegrityCheck()
...
+CAccount* CAccountManager::Get(const char* szName, const char* szPassword, bool bCaseSensitive)
 {
     if (szName && szName[0])
     {
         std::vector<CAccount*> results;
-        m_List.FindAccountMatches(&results, szName, true);
+        m_List.FindAccountMatches(&results, szName, bCaseSensitive);
+        if (!bCaseSensitive)
+        {
+            CAccount* firstMatchAccount = nullptr;
```
Comment (botder) [inline Server/mods/deathmatch/logic/CAccountManager.cpp]: Prefix this variable with a **p**: `CAccount* pFirstMatchAccount = nullptr;`

**Review by botder — CHANGES_REQUESTED** [review]

**Review by qaisjp — APPROVED** [review]
Please address @botder's review, otherwise, looks good.

**Pull request merged**